docs/assemblies/proc_creating-a-custom-service.adoc

```diff
@@ -52,16 +52,16 @@ metadata:
 spec:
   label: dataplane-deployment-custom-service
   play: |
-    hosts: all
-    tasks:
-      - name: Hello World!
-        shell: "echo Hello World!"
-        register: output
-      - name: Show output
-        debug:
-          msg: "{{ output.stdout }}"
-      - name: Hello World role
-        import_role: hello_world
+    - hosts: all
```

Comment:
`Play` is actually a dict[1] and not a list where as `Playbook` is a list of plays. I think we should change the webhook validation logic.

[1] https://github.com/ansible/ansible/blob/devel/test/units/playbook/test_play.py#L48

Reply:
Ok, if we go down that path, then the docs change we need is this line:
https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/docs/assemblies/proc_creating-a-custom-service.adoc?plain=1#L99

Reply:
agree, I think this should remain a dict.

Reply:
Ok, so we're removing the Hyphen from L99 then for this change yeah?

Updated this PR accordingly.

Reply:
So, here's a test service:
```
$ cat adhoc-service.yaml
      apiVersion: dataplane.openstack.org/v1beta1
      kind: OpenStackDataPlaneService
      metadata:
        name: random-adhoc
      spec:
        play: |
            hosts: all
            become: true
            tasks:
            - name: Test task
              ansible.builtin.debug:
                msg: Success!
```
Here's the deployment to deploy it:
```
$ cat test-adhoc-deployment.yaml
      apiVersion: dataplane.openstack.org/v1beta1
      kind: OpenStackDataPlaneDeployment
      metadata:
        name: edpm-adhoc
        namespace: openstack
      spec:
        nodeSets:
          - openstack-edpm-ipam
        servicesOverride:
          - random-adhoc
```

Results in:
```
$ oc get osdpd edpm-adhoc -o yaml | yq '.status.conditions[0].message'
Deployment error occurred nodeSet: openstack-edpm-ipam error: admission webhook "vopenstackansibleee.kb.io" denied the request: OpenStakAnsibleEE.ansibleee.openstack.org "random-adhoc-edpm-adhoc-openstack-edpm-ipam" is invalid: spec.play: Invalid value: "hosts: all\nbecome: true\ntasks:\n- name: Test task\n  ansible.builtin.debug:\n    msg: Success!\n": error checking sanity of an inline play: hosts: all
become: true
tasks:
- name: Test task
  ansible.builtin.debug:
    msg: Success!
 Key: '' Error:Field validation for '' failed on the 'play' tag
```


So before we merge this change to keep the dict, we'll need to look into this validation some more.

Reply:
And I'll dump a working one here too so that we have a full reproducer in one place:

This one with the Hyphen:

Service:
```
$ cat adhoc-service-hyphen.yaml
      apiVersion: dataplane.openstack.org/v1beta1
      kind: OpenStackDataPlaneService
      metadata:
        name: random-adhoc-hyphen
      spec:
        play: |
          - hosts: all
            become: true
            tasks:
            - name: Test task
              ansible.builtin.debug:
                msg: Success!
```

Deployment:
```
$ cat test-adhoc-hyphen-deployment.yaml
      apiVersion: dataplane.openstack.org/v1beta1
      kind: OpenStackDataPlaneDeployment
      metadata:
        name: edpm-adhoc-hyphen
        namespace: openstack
      spec:
        nodeSets:
          - openstack-edpm-ipam
        servicesOverride:
          - random-adhoc-hyphen
```

```
$ oc get osdpd edpm-adhoc-hyphen -o yaml | yq .status.conditions
- lastTransitionTime: "2024-05-23T01:04:19Z"
  message: Deployment in progress
  reason: Requested
  severity: Info
  status: "False"
  type: Ready
- lastTransitionTime: "2024-05-23T01:04:19Z"
  message: Deployment in progress
  reason: Requested
  severity: Info
  status: "False"
  type: DeploymentReady
- lastTransitionTime: "2024-05-23T01:04:19Z"
  message: Setup complete
  reason: Ready
  status: "True"
  type: InputReady
```

Reply:
We need to change the terminology in our description of the Play object:
https://github.com/openstack-k8s-operators/dataplane-operator/pull/890

And this is for the ansibleee validation:
https://github.com/openstack-k8s-operators/openstack-ansibleee-operator/pull/362

Reply:
you're right...I stand corrected. Spec.Play needs to be a list as it gets echo'd directly to /runner/project/playbook.yaml, so it's treated as the playbook contents, not a play. 

So...I'd suggest renaming the Spec field. As we already have Spec.Playbook for specifying the playbook name, we could change Spec.Play to Spec.PlaybookContents.